### ISSUE TITLE
Array to string conversion in TsvFormatter.php on line 35

### ISSUE THREAD

**Issue by MPParsley** [drush-ops/drush#3721]
**Describe the bug**
`drush locale:check` gives this notice: `Array to string conversion in TsvFormatter.php on line 35`

**To Reproduce**
`drush locale:check`

**Expected behavior**
No notice should appear.

**Actual behavior**
A notice: `Array to string conversion in TsvFormatter.php on line 35`

**Workaround**
-

### System Configuration
| Q               | A
| --------------- | ---
| Drush version?  | 9.4
| Drupal version? | 8.6
| PHP version     | 7.2.10
| OS?             | Mac

**Additional information**
At the end of the script this message appears:
`Array`

**Comment by greg-1-anderson** [comment]
What output do you get with `drush locale:check --format=yaml`?  Whether or not you get this warning is likely dependent on the result returned by the command, which may vary based on the state of your Drupal installation.

**Comment by MPParsley** [comment]
I get this error:
`  The "--format" option does not exist.  `

Possibly not related but note that I also encountered #3722 and https://www.drupal.org/project/drupal/issues/3005958.

**Comment by greg-1-anderson** [comment]
`locale:check` does not return any data, so it should not be running the format manager / tsv formatter.

When I run this command, I get no output and no warning. Could you provide steps to reproduce from a fresh install?

**Comment by MPParsley** [comment]
It might be related to the PHP version.
I'm on `7.2.10` and a colleague has the same project setup on `7.2.3` and doesn't have this issue.

The array that is sent to the format manager is a list of all modules and the manager is unsuccessfully trying to format it as a table.

I'll document the steps.

**Comment by greg-1-anderson** [comment]
You're using [locale:check](https://github.com/drush-ops/drush/blob/9.4.0/src/Drupal/Commands/core/LocaleCommands.php#L45)? I'm also using php 7.2.10. local:check does not use the formatter manager, so I'm a bit perplexed as to why this is going through the tsv formatter.

**Comment by MPParsley** [comment]
Yes, I'm using `drush locale:check`. See [this screencast](https://www.dropbox.com/s/m4gjk3kmdkfshpd/recording.mov?dl=0).

**Comment by greg-1-anderson** [comment]
writeUsingFormatter should not be called unless this function returns true:
```
    protected function dataCanBeFormatted($structuredOutput)
    {
        if (!isset($this->formatterManager)) {
            return false;
        }
        return
            is_object($structuredOutput) ||
            is_array($structuredOutput);
    }
```
However, `$structuredOutput` should be null, since `locale:check` does not produce any output.

In theory, anyway. I'll have to reproduce this locally before I can make any progress.

**Comment by greg-1-anderson** [comment]
I could potentially work around the symptom by making tsv detect when it is passed data that cannot be structured as a table. However, it would be good to know where that data is being provided from, since the main command handler for `locale:check` certainly does not return anything. Some hook somewhere, but where and why?

**Comment by MPParsley** [comment]
The data comes from here in `locale_translate_batch_import()`:
         $context['results']['files'][$file->uri] = $file;

**Comment by MPParsley** [comment]
So it turns out to be a Drupal core issue. See https://www.drupal.org/project/drupal/issues/3006084

**Comment by greg-1-anderson** [comment]
Hm, I do wonder, though, why `$context['results']` ends up getting printed by Drush. Maybe we have some hook I don't know about that's trying to be helpful by injecting the batch results into the command output?

**Comment by greg-1-anderson** [comment]
Couldn't find anything like that, although I did find `MessengerCommands::log()` that replicates messenger logs to the console.

**Comment by greg-1-anderson** [comment]
UpdateDBCommands.php looks at `$context['results']` but that shouldn't be at work here.

**Comment by MPParsley** [comment]
Indeed, we may have to reopen this issue.

In batch.inc, the output of the finish callback bubbles all the way up to `drush_batch_command` which is called in `BatchCommands`:

```
class BatchCommands extends DrushCommands
{

    /**
     * Process operations in the specified batch set.
     *
     * @command batch:process
     * @aliases batch-process
     * @param $batch_id The batch id that will be processed.
     * @hidden
     */
    public function process($batch_id)
    {
        return drush_batch_command($batch_id);
    }
}

```

Now CommandProcessor::dataCanBeFormatted() returns true when it's an array but in the case of locale it is a nested array.
```
    protected function dataCanBeFormatted($structuredOutput)
    {
        if (!isset($this->formatterManager)) {
            return false;
        }
        return
            is_object($structuredOutput) ||
            is_array($structuredOutput);
    }
```

In `locale.batch.inc` results is set as a nested array:
```
  // Provide user feedback and record success or failure for reporting at the
  // end of the batch.
  if ($options['finish_feedback'] && $checked) {
    $context['results']['files'][] = $source->name;
  }
  if ($failure && !$checked) {
    $context['results']['failed_files'][] = $source->name;
  }
```

So in CommandProcessor::handleResults() the output of results['files'][] & results['failed_files'][] is written:

```
    public function handleResults(OutputInterface $output, $names, $result, CommandData $commandData)
    {
        $statusCodeDispatcher = new StatusDeterminerHookDispatcher($this->hookManager(), $names);
        $status = $statusCodeDispatcher->determineStatusCode($result);
        // If the result is an integer and no separate status code was provided, then use the result as the status and do no output.
        if (is_integer($result) && !isset($status)) {
            return $result;
        }
        $status = $this->interpretStatusCode($status);

        // Get the structured output, the output stream and the formatter
        $extractDispatcher = new ExtracterHookDispatcher($this->hookManager(), $names);
        $structuredOutput = $extractDispatcher->extractOutput($result);
        $output = $this->chooseOutputStream($output, $status);
        if ($status != 0) {
            return $this->writeErrorMessage($output, $status, $structuredOutput, $result);
        }
        if ($this->dataCanBeFormatted($structuredOutput) && isset($this->formatterManager)) {
            return $this->writeUsingFormatter($output, $structuredOutput, $commandData);
        }
        return $this->writeCommandOutput($output, $structuredOutput);
    }
```

Personally I don't think it's a good idea to use the results array to store success/failure data, the API describes it as a set of results. That's why I opened #3006084.
On the other hand, if there is output, it can be a nested array and it should probably be handled?

**Comment by greg-1-anderson** [comment]
So, the real bug here is that `batch:process` does not declare that it returns any data, but then it goes ahead and returns something anyway. The format manager presumes this means that the command meant that it returns a STRING. If a command declares that it returns a string, but instead returns an array, then it falls back to using tsv for the output. This might sound like a really bad decision; however, tsv is actually the best answer in instances where the data type is downgraded, e.g. when the user includes the `--field` option.

While we could certainly improve the fallback behavior of the output-formatter library, I think that the real solution would be to have `batch:process` return its data wrapped in UnstructuredListData (n.b. requires https://github.com/consolidation/output-formatters/pull/69, which has not been merged yet), and declare its default data type to be `yaml`.

**Comment by MPParsley** [comment]
@greg-1-anderson, I agree. Should we reopen this issue or how do you suggest we proceed?

**Comment by greg-1-anderson** [comment]
There is now a release of output formatters that contains the UnstructuredData and UnstructuredListData types, and it is in use on the `master` branch of Drush. If you want to keep working on this, I would recommend making a new PR that sets the data type and default format of `batch:process`.

The main question in that PR is going to be whether the resulting output is then rational. Is the information present useful and interesting to the user?  If not, we might want to consider making batch process return its result to backend invoke without printing the result.

**Comment by greg-1-anderson** [comment]
I think that the best way to solve this is to have `batch:process` declare that it returns UnstructuredListData, and its default format is NONE.  I'll make a new release of output-formatters that includes a NONE format type.

**Comment by greg-1-anderson** [comment]
https://github.com/consolidation/output-formatters/pull/71

**Comment by MPParsley** [comment]
Thanks, https://github.com/drush-ops/drush/pull/3750 has fixed both issues:

- Notice: Array to string conversion in TsvFormatter.php
- This output: "Array   Array   Array"

**Comment by MPParsley** [comment]
@greg-1-anderson seems like #3750 never got merged so this issue remains.

**Comment by greg-1-anderson** [comment]
In theory it was merged as part of #3758. Could you compare #3750 with master and make a PR to re-fix?

**Comment by MPParsley** [comment]
@greg-1-anderson seems like these are the only changes:

-    "consolidation/output-formatters": "^3.3.1",
+    "consolidation/output-formatters": "^3.4",

-    public function process($batch_id, $options = ['format' => 'json'])
+    public function process($batch_id, $options = ['format' => 'null'])
